In [29]:
# %matplotlib widget

from __future__ import annotations


import matplotlib.colors as mcolors
from test_utilities import test, TestDict


COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 7: Laboratories ---</h2><p>You thank the cephalopods for the help and exit the trash compactor, finding yourself in the <a href="/2024/day/6">familiar</a> <a href="/2018/day/4">halls</a> of a North Pole research wing.</p>
<p>Based on the large sign that says "teleporter hub", they seem to be researching <em>teleportation</em>; you can't help but try it for yourself and step onto the large yellow teleporter pad.</p>
<p>Suddenly, you find yourself in an unfamiliar room! The room has no doors; the only way out is the teleporter. Unfortunately, the teleporter seems to be leaking <a href="https://en.wikipedia.org/wiki/Magic_smoke" target="_blank">magic smoke</a>.</p>
<p>Since this is a teleporter lab, there are lots of spare parts, manuals, and diagnostic equipment lying around. After connecting one of the diagnostic tools, it helpfully displays error code <code>0H-N0</code>, which apparently means that there's an issue with one of the <em>tachyon manifolds</em>.</p>
<p>You quickly locate a diagram of the tachyon manifold (your puzzle input). A tachyon beam enters the manifold at the location marked <code>S</code>; tachyon beams always move <em>downward</em>. Tachyon beams pass freely through empty space (<code>.</code>). However, if a tachyon beam encounters a splitter (<code>^</code>), the beam is stopped; instead, a new tachyon beam continues from the immediate left and from the immediate right of the splitter.</p>
<p>For example:</p>
<pre><code>.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
</code></pre>
<p>In this example, the incoming tachyon beam (<code>|</code>) extends downward from <code>S</code> until it reaches the first splitter:</p>
<pre><code>.......S.......
.......|.......
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
</code></pre>
<p>At that point, the original beam stops, and two new beams are emitted from the splitter:</p>
<pre><code>.......S.......
.......|.......
......|^|......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
</code></pre>
<p>Those beams continue downward until they reach more splitters:</p>
<pre><code>.......S.......
.......|.......
......|^|......
......|.|......
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
</code></pre>
<p>At this point, the two splitters create a total of only <em>three</em> tachyon beams, since they are both dumping tachyons into the same place between them:</p>
<pre><code>.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
</code></pre>
<p>This process continues until all of the tachyon beams reach a splitter or exit the manifold:</p>
<pre><code>.......S.......
.......|.......
......|^|......
......|.|......
.....|^|^|.....
.....|.|.|.....
....|^|^|^|....
....|.|.|.|....
...|^|^|||^|...
...|.|.|||.|...
..|^|^|||^|^|..
..|.|.|||.|.|..
.|^|||^||.||^|.
.|.|||.||.||.|.
|^|^|^|^|^|||^|
|.|.|.|.|.|||.|
</code></pre>
<p>To repair the teleporter, you first need to understand the beam-splitting properties of the tachyon manifold. In this example, a tachyon beam is split a total of <em><code>21</code></em> times.</p>
<p>Analyze your manifold diagram. <em>How many times will the beam be split?</em></p>
</article>


In [30]:
from collections import deque
from pprint import pprint
import re
from util import Str

tests: list[TestDict] = [
    {
        "name": "Example",
        "s": """
            .......S.......
            ...............
            .......^.......
            ...............
            ......^.^......
            ...............
            .....^.^.^.....
            ...............
            ....^.^...^....
            ...............
            ...^.^...^.^...
            ...............
            ..^...^.....^..
            ...............
            .^.^.^.^.^...^.
            ...............
        """,
        "expected": 21,
    },
]


class Diagram(Str):
    def __init__(self, s: str) -> None:
        self.grid = [re.findall(r"[\.^S]", l) for l in s.strip().splitlines()]
        self.rows, self.cols = len(self.grid), len(self.grid[0])
        self.S: tuple[int, int] = 0, self.grid[0].index("S")

    def split_the_beam(self) -> int:
        count = 0
        queue = deque([self.S])
        seen: set[tuple[int, int]] = set()

        while queue:
            row, col = queue.popleft()

            if not (0 <= row < self.rows and 0 <= col <= self.cols):
                continue

            if (row, col) in seen:
                continue

            seen.add((row, col))

            if self.grid[row][col] == "^":
                count += 1
                queue.append((row, col - 1))
                queue.append((row, col + 1))
            else:
                queue.append((row + 1, col))

        return count


@test(tests=tests)
def part_I(s: str) -> int:
    d = Diagram(s)
    return d.split_the_beam()


Test Example passed, for part_I.
Success


In [31]:
with open("../input/day7.txt") as f:
    puzzle = f.read()

print(f"Part I: {part_I(puzzle)}")

Part I: 1600


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1600</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>With your analysis of the manifold complete, you begin fixing the teleporter. However, as you open the side of the teleporter to replace the broken manifold, you are surprised to discover that it isn't a classical tachyon manifold - it's a <em><span title="Please disregard the wave interference patterns that would arise from the wave-particle duality of individual tachyon particles while repairing the manifold.">quantum</span> tachyon manifold</em>.</p>
<p>With a quantum tachyon manifold, only a <em>single tachyon particle</em> is sent through the manifold. A tachyon particle takes <em>both</em> the left and right path of each splitter encountered.</p>
<p>Since this is impossible, the manual recommends the many-worlds interpretation of quantum tachyon splitting: each time a particle reaches a splitter, it's actually <em>time itself</em> which splits. In one timeline, the particle went left, and in the other timeline, the particle went right.</p>
<p>To fix the manifold, what you really need to know is the <em>number of timelines</em> active after a single particle completes all of its possible journeys through the manifold.</p>
<p>In the above example, there are many timelines. For instance, there's the timeline where the particle always went left:</p>
<pre><code>.......S.......
.......|.......
......|^.......
......|........
.....|^.^......
.....|.........
....|^.^.^.....
....|..........
...|^.^...^....
...|...........
..|^.^...^.^...
..|............
.|^...^.....^..
.|.............
|^.^.^.^.^...^.
|..............
</code></pre>
<p>Or, there's the timeline where the particle alternated going left and right at each splitter:</p>
<pre><code>.......S.......
.......|.......
......|^.......
......|........
......^|^......
.......|.......
.....^|^.^.....
......|........
....^.^|..^....
.......|.......
...^.^.|.^.^...
.......|.......
..^...^|....^..
.......|.......
.^.^.^|^.^...^.
......|........
</code></pre>
<p>Or, there's the timeline where the particle ends up at the same point as the alternating timeline, but takes a totally different path to get there:</p>
<pre><code>.......S.......
.......|.......
......|^.......
......|........
.....|^.^......
.....|.........
....|^.^.^.....
....|..........
....^|^...^....
.....|.........
...^.^|..^.^...
......|........
..^..|^.....^..
.....|.........
.^.^.^|^.^...^.
......|........
</code></pre>
<p>In this example, in total, the particle ends up on <em><code>40</code></em> different timelines.</p>
<p>Apply the many-worlds interpretation of quantum tachyon splitting to your manifold diagram. <em>In total, how many different timelines would a single tachyon particle end up on?</em></p>
</article>


In [32]:
from functools import cache
from PrettyPrint import PrettyPrintTree

tests: list[TestDict] = [
    {
        "name": "Example",
        "s": """
            .......S.......
            ...............
            .......^.......
            ...............
            ......^.^......
            ...............
            .....^.^.^.....
            ...............
            ....^.^...^....
            ...............
            ...^.^...^.^...
            ...............
            ..^...^.....^..
            ...............
            .^.^.^.^.^...^.
            ...............
        """,
        "expected": 40,
    },
]


class DiagramII(Diagram):
    def __init__(self, s: str) -> None:
        super().__init__(s)
        self.graph = self._make_graph()

    def _make_graph(self) -> dict[tuple[int, int], list[tuple[int, int]]]:
        graph = {self.S: []}
        queue = deque([(*self.S, *self.S)])

        while queue:
            row, col, parent_row, parent_col = queue.popleft()

            if not (0 <= row < self.rows and 0 <= col <= self.cols):
                graph[(parent_row, parent_col)].append((row, col))
                continue

            if (parent_row, parent_col) != (row, col) and (row, col) in graph:
                graph[(parent_row, parent_col)].append((row, col))
                continue

            if self.grid[row][col] == "^":
                graph[(row, col)] = []
                graph[(parent_row, parent_col)].append((row, col))
                queue.append((row, col - 1, row, col))
                queue.append((row, col + 1, row, col))
            else:
                queue.append((row + 1, col, parent_row, parent_col))

        return graph

    def print_graph(self) -> None:
        pt = PrettyPrintTree(lambda x: self.graph.get(x, []), lambda x: str(x))  # type: ignore
        pt(self.S)  # type: ignore

    def timelines(self) -> int:
        @cache
        def dfs(node: tuple[int, int]) -> int:
            if node not in self.graph:
                return 1

            return sum(dfs(neigbor) for neigbor in self.graph[node])

        return dfs(self.S)


@test(tests=tests)
def part_II(s: str) -> int:
    d = DiagramII(s)
    return d.timelines()


Test Example passed, for part_II.
Success


<link href="style.css" rel="stylesheet"></link>


In [33]:
print(f"Part II: {part_II(puzzle)}")

Part II: 8632253783011


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>8632253783011</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
